scripts/builtin/hmm.dml

```diff
@@ -0,0 +1,258 @@
+    from 1 to some max_states or 1 to log(n_timesteps), depending on whichever is greater.
+
+    Also the break condition is: If the likelihood decreases with increase in number of total states, 
+    break and take the paremeters of the last iteration as the optimal one.
+    */
+    
+    max_states = 6
+    T = ncol(X)
+    
+    if (max_states < log(T)){
+        max_states = ceil(log(T))
+    }
+    
+    search = TRUE
+    nr_states = 2
+    seed = 42
+
+    while (search) {
+        [A_new, B_new, ip_new, curr_ll, seed] = baum_welch(X, nr_states, nr_outputs, 20, seed)
+        seed = seed+1
+        if (nr_states == 2) {
+            prev_ll = -1
+        }
+        if (curr_ll < prev_ll) {
+            search = FALSE
+        }
+        else {
+            A = A_new
+            B = B_new
+            ip = ip_new
+        }
+        prev_ll = curr_ll
+        nr_states = nr_states+1
+
+        if (nr_states > max_states) {
+            search = FALSE
+        }
+    }
+}
+
+col_normalize = function(Matrix[Double] m) return (Matrix[Double] m_norm) {
+	m_norm = m / rowSums(m)
+}
+
+unique_vals = function (Matrix[Double] X) return (Integer l) {
+    # group and count duplicates
+    I = aggregate (target = X, groups = X[,1], fn = "count")
+    # select groups
+    res = removeEmpty (target = seq (1, max (X[,1])), margin = "rows", select = (I != 0))
+    l = length(res)
+}
+
+baum_welch = function( Matrix[Double] X, Integer nr_states, Integer nr_outputs, Integer IT, Integer seed) return (Matrix[Double] A, Matrix[Double] B, Matrix[Double] ip, Double likelihood, Integer seed)
+{
+    #initialize state transition and emmission matrices uniformly
+    A = col_normalize(matrix(1/nr_states, rows=nr_states, cols=nr_states) + rand(rows=nr_states, cols=nr_states, seed=seed))
+    seed = seed+1
+    B = col_normalize(matrix(1/nr_outputs,rows=nr_states, cols=nr_outputs) + rand(rows=nr_states, cols=nr_outputs, seed=seed))
+    seed = seed+1
+    ip = matrix(1/nr_states, rows=nr_states, cols=1) + rand(rows=nr_states, cols=1, seed=seed)
+    seed = seed+1
+
+    likelihood = 0
+    T = ncol(X)
+    
+    for (i in 1:IT) {
+        alpha = forward(X, A, B, ip)
+        beta = backward(X, A, B)
+        
+        gamma = calculate_gamma(alpha, beta)
+        eta = calculate_eta(X, alpha, beta, A, B)
+
+        ip = gamma[, 1]
+        A = calculate_A(eta, gamma, nr_states)
+        B = calculate_B(gamma, X, nr_states, nr_outputs)
+        
+        #likelhood using forward method
+        likelihood = sum(gamma[, T] * alpha[, T])
+    }
+}
+
+calculate_A = function(Matrix[Double] eta, Matrix[Double] gamma, Integer nr_states) return (Matrix[Double] A) {
+    A = rand(rows=nr_states, cols=nr_states)
+    T = ncol(eta)
+
+    parfor (i in 1:nr_states) {
+        parfor (j in 1:nr_states) {
+            eta_id = (i-1) * nr_states + j
+            num_ij = sum(eta[eta_id, 1:T-1])
```

Comment:
1:T-1 takes all columns except the last one. Is this the intended behavior or should it take the entire row?